library/src/main/java/com/isapp/chips/ChipsView.kt

```diff
@@ -0,0 +1,205 @@
+          }
+          else if(chip.deletable && !chip.icon) {
+            (rect.width() + context.dip(55)).toFloat()
+          }
+          else if(!chip.deletable && chip.icon) {
+            (rect.width() + context.dip(60)).toFloat()
+          }
+          else {
+            (rect.width() + context.dip(35)).toFloat()
+          }
+
+          val widthPerSpan = parentWidth.toFloat() / spanCount.toFloat()
+          return Math.ceil(childWidth / widthPerSpan.toDouble()).toInt()
+        }
+      }
+
+      spanSizeLookup.isSpanIndexCacheEnabled = true
+
+      addItemDecoration(GridSpacingItemDecoration(10))
+    }
+    adapter = adapter?.let(::ChipsAdapter) ?: ChipsAdapter()
+    swapAdapter(adapter, true)
+  }
+
+  fun useHorizontalScrollingLayout() = synchronized(this) {
+    layoutManager = LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false)
+    adapter = adapter?.let(::ChipsAdapter) ?: ChipsAdapter()
+    swapAdapter(adapter, true)
+  }
+}
+
+data class Chip(val data: Any, val text: String = data.toString(), val deletable: Boolean = false, val icon: Boolean = false)
+
+private class ChipsAdapter() : RecyclerView.Adapter<ChipsViewHolder>() {
+  companion object {
+    const val JUST_TEXT = 0
+    const val DELETABLE = 1
+    const val ICON = 2
+    const val DELETABLE_ICON = 3
+  }
+
+  constructor(adapter: ChipsAdapter) : this() {
+    chips.addAll(adapter.chips)
+    textAppearance = adapter.textAppearance
+    listener = adapter.listener
+  }
+  
+  internal val chips: MutableList<Chip> = ArrayList()
+  internal var textAppearance: Int = 0
+  internal var listener: ChipsListener? = null
+
+  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ChipsViewHolder {
+    return ChipsViewHolder(
+        when(viewType) {
+          JUST_TEXT -> LayoutInflater.from(parent.context).inflate(R.layout.chip, parent, false)
+          DELETABLE -> LayoutInflater.from(parent.context).inflate(R.layout.deletable_chip, parent, false)
+          ICON ->  LayoutInflater.from(parent.context).inflate(R.layout.icon_chip, parent, false)
+          else -> LayoutInflater.from(parent.context).inflate(R.layout.deletable_icon_chip, parent, false)
+        }
+    ).apply {
+      if(textAppearance > 0) {
+        TextViewCompat.setTextAppearance(text, textAppearance)
+      }
+    }
+  }
+
+  override fun onBindViewHolder(holder: ChipsViewHolder, position: Int) {
+    val chip = chips[position]
+    holder.itemView.setOnClickListener {
+      listener?.onChipClicked(chips[holder.adapterPosition])
+    }
+    holder.text.text = chip.text
+    holder.icon?.let {
+      listener?.onLoadIcon(chip, it)
+    }
+    holder.delete?.setOnClickListener {
+      listener?.onChipDeleted(chips[holder.adapterPosition])
+    }
+  }
+
+  override fun getItemCount() = chips.size
+
+  override fun getItemViewType(position: Int): Int {
+    val chip = chips[position]
+    return if(!chip.deletable and chip.icon) {
```

Comment:
"Not deletable and has icon" = `JUST_TEXT`?  Looks like it should be `ICON`?  Looks like maybe all these are mixed up?